src/storage/src/upsert/types.rs

```diff
@@ -189,6 +189,7 @@ pub enum Value<O> {
 /// A value as produced during consolidation of a snapshot.
 #[derive(Clone, Default, serde::Serialize, serde::Deserialize, Debug)]
 pub struct Snapshotting {
+    #[serde(with = "serde_bytes")]
```

Comment:
This is a benchmark of all competing libraries https://github.com/djkoloski/rust_serialization_benchmark

Reply:
oh good call out -- I'll see if I can switch over from `bincode` to `bitcode` and then we wouldn't need `serde_bytes` either

Reply:
Turns out that the version of the `bitcode` with the crazy performance requires that all types being encoded implement the `bitcode` Encode/Decode traits. Since I can't add those derives to the underlying `Row` and `CompactBytes` structs we can't use `bitcode` on these values directly.

`bitcode` does have support for a different encoding scheme on types that implement serde  `Serialize`/`Deserialize` but it's quite a bit slower than the default scheme, and when I tried to get it working on our types it fails to deserialize the `CompactBytes` field inside our `Row` struct for some reason I can't figure out. With this serde feature it also doesn't have any ability to reuse buffers so would likely additionally be slower because of the additional allocations.

I think it's worth watching the `bitcode` crate to see if it adds better support for encoding specific fields with serde in the future, but as of now that's not possible.

Reply:
bummer. `Row` is just a byte slice internally that we can get access to from `Row::data`, although it's marked as debugging only. It doesn't sound unreasonable to add the encode/decode derive on it though if it makes upsert much faster. Thanks for looking into it!

Reply:
At it was a byte-slice but now uses https://github.com/ParkMyCar/compact_bytes -- which I just realized is written by @ParkMyCar ! Maybe we can just implement the `bitcode` derives on it then

Reply:
Oh, I see. `bitcode::Encode` is not a trait that we can manually implement. What I had in mind was that we could implement it ourselves and give it the slice of bytes returned by `Row::data` for bitcode to write down.

Reply:
Yes exactly.
I'm going to merge as-is and potentially come back to the `bitcode` stuff later.

For posterity here are the two in-progress branches:
- bitcode Encode/Decode: https://github.com/rjobanp/materialize/commit/b1b2a186874e36265ebf9262d5d4138d05439154 - blocked by not being able to derive Encode/Decode on Row/CompactBytes
- bitcode w/ serde feature: https://github.com/rjobanp/materialize/commit/02eff3fccc885d5d77db319a6307c910dfb17a51 - blocked by a failure on bitcode::deserialize:
```
cargo test --package mz-storage --lib -- upsert::types::tests::test_merge_update
...
thread 'upsert::types::tests::test_merge_update' panicked at src/storage/src/upsert/types.rs:410:58:
called `Result::unwrap()` on an `Err` value: Error("invalid type: byte array, expected bytes")
```

Reply:
Got around to figuring out that error above to potentially enable using `bitcode` with the serde feature - it requires an update of the `compact_bytes` package https://github.com/ParkMyCar/compact_bytes/pull/4

Reply:
Following up to this -- I got the `bitcode` with `serde` method working in this commit https://github.com/rjobanp/materialize/commit/cd629240ff17199f995cccbdc352819cd78ed9b7

however performance testing with the upsert `load-gen` test doesn't show a significant gain with the merge operator use.

On main:
```
Scenario default took [datetime.timedelta(seconds=12, microseconds=345000)] ms
Scenario default with RocksDB Merge Operator took [datetime.timedelta(seconds=10, microseconds=724000)] ms
```

On this branch:
```
Scenario default took [datetime.timedelta(seconds=11, microseconds=526000)] ms
Scenario default with RocksDB Merge Operator took [datetime.timedelta(seconds=10, microseconds=56000)] ms
```

The other downside with this approach is that there is no way to reuse allocations between serialization calls (no `serialize_into` method) when using the `serde` feature.

I also tried directly deriving the `bitcode` `Encode` / `Decode` traits on the `CompactBytes` type but since it's a union type that wasn't possible either. So I think the `bitcode` path is a dead-end, overall.

Reply:
Thanks for digging into this and for the great write up!